File: lesson-7/java/Vehicle.java
/*
 * This is the vehicle class
 *
 * @author Mohamad Tanbari
 * @version 1.0.0
 * @since 2024-11-04
 */

/**
 * The vehicle class.
 */
public class Vehicle {
    /**
     * The color of the vehicle.
     */
    private String color;

    /**
     * The current speed of the vehicle.
     */
    private int speed;

    /**
     * The max speed of the vehicle.
     */
    private int maxSpeed;

    /**
     * Constructor.
     *
     * @param color - The color you want the vehicle to start with.
     * @param maxSpeed - The max speed the vehicle can reach.
     */
    public Vehicle(
        String color,
        int maxSpeed) {
        // Set the properties
        this.color = color;
        this.maxSpeed = maxSpeed;
        this.speed = 0;
    }

    /**
     * The getter for the color property.
     *
     * @return A string containing the color of the vehicle.
     */
    public String getColor() {
        return this.color;
    }

    /**
     * The setter for the color property.
     *
     * @param newColor - THe new color to set the vehicle to.
     */
    public void setColor(String newColor) {
        this.color = newColor;
    }

    /**
     * The getter for the speed property.
     *
     * @return An integer of the current speed of the vehicle.
     */
    public int getSpeed() {
        return this.speed;
    }

    /**
     * The setter for the speed property.
     *
     * @param newSpeed - The new speed to set the vehicle to.
     */
    public void setSpeed(int newSpeed) {
      if (newSpeed < 0) {
          this.speed = 0;
      } else if (newSpeed > this.maxSpeed) {
          this.speed = this.maxSpeed;
      } else {
          this.speed = newSpeed;
      }
    }

    /**
     * The accelerate method to increase the speed of the vehicle.
     *
     * @param accelerationPower - The intensity of the acceleration.
     * @param accelerationTime - The time the vehicle accelerates for.
     */
    public void accelerate(int accelerationPower, int accelerationTime) {
        this.speed = (accelerationPower * accelerationTime) + this.speed;

        if (this.speed > this.maxSpeed) {
            this.speed = this.maxSpeed;
        }
    }

    /**
     * The brake method to slow down the vehicle.
     *
     * @param brakePower - The intensity of the brakes.
     * @param brakeTime - The time the vehicle is braking for.
     */
    public void brake(int brakePower, int brakeTime) {
        this.speed = this.speed - (brakePower * brakeTime);

        // Ensure speed does not fall below zero
        if (this.speed < 0) {
            this.speed = 0;
        }
    }
}

File: lesson-7/ts/Vehicle.ts
/**
 * This is the Vehicle class
 *
 * By: Mohamad Tanbari
 * Version: 1.0
 * Since: 2024-10-31
 */

export default class Vehicle {
  /**
   * The color of the vehicle.
   */
  private privateColour: string

  /**
   * The current speed of the vehicle.
   */
  private privateSpeed: number = 0

  /**
   * The max speed of the Vehicle.
   */
  private readonly _maxSpeed: number

  /**
   * The constructor to change the values of the variables.
   */
  constructor (colour: string, maxSpeed: number) {
    this.privateColour = colour
    this._maxSpeed = maxSpeed
  }

  /**
   * The getter for the colour property.
   *
   * @return {string} colour - The current colour of the vehicle.
   */
  public get colour (): string {
    return this.privateColour
  }

  /**
   * The setter for the colour property.
   *
   * @param {string} newColour - The new color to set the vehicle as.
   */
  public set colour (newColour: string) {
    this.privateColour = newColour
  }

  /**
   * The getter for the speed of the vehicle.
   *
   * @return {number} speed - The current speed of the vehicle.
   */
  public get speed (): number {
    return this.privateSpeed
  }

  /**
   * The setter for the speed property.
   *
   * @param {number} newSpeed - The new speed to set the vehicle to.
   */
  public set speed (newSpeed: number) {
    if (newSpeed < 0) {
      this.privateSpeed = 0
    } else {
      this.privateSpeed = newSpeed
    }
  }

  /**
   * The getter for the max speed property.
   *
   * @return {number} The max speed of the vehicle/
   */
  public get maxSpeed (): number {
    return this._maxSpeed
  }

  /**
   * The accelerate method to increase the speed of the vehicle.
   *
   * @param {number} accelerationPower - The intensity of the acceleration.
   * @param {number} accelerationTime - The time the vehicle accelerates for.
   */
  public accelerate (accelerationPower: number,
    accelerationTime: number): void {
    this.privateSpeed = (accelerationPower * accelerationTime) + this.privateSpeed

    // Ensure that the vehicle does not accelerate beyond max speed
    if (this.privateSpeed > this._maxSpeed) {
      this.privateSpeed = this._maxSpeed
    }
  }

  /**
   * The break method to slow down the vehicle.
   *
   * @param {number} brakePower - The intensity of the brakes.
   */
  public brake (brakePower: number, brakeTime: number): void {
    this.privateSpeed = this.privateSpeed - (brakePower * brakeTime)

    // Ensure speed does not fall below zero
    if (this.privateSpeed < 0) {
      this.privateSpeed = 0
    }
  }
}
